operator/cnp_event.go

```diff
@@ -101,3 +130,179 @@ func enableCNPWatcher() error {
+	node string
+	*cilium_v2.CiliumNetworkPolicyNodeStatus
+}
+
+var controllers = controller.NewManager()
+
+func extractFieldsFromKey(key string) (namespace, name, node string, err error) {
+	withoutPrefix := strings.TrimLeft(key, k8s.CNPStatusesPath)
+	// result is now namespace/name/node
+	splitStr := strings.Split(withoutPrefix, "/")
+	if len(splitStr) != 3 {
+		err = fmt.Errorf("could not parse key: %s", key)
+		return
+	}
+	return splitStr[0], splitStr[1], splitStr[2], nil
+
+}
+
+func getUpdatedCNPFromStore(ciliumStore cache.Store, nameNamespace string) (*types.SlimCNP, error) {
+	serverRuleStore, exists, err := ciliumStore.GetByKey(nameNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find v2.CiliumNetworkPolicy in local cache: %s", err)
+	}
+	if !exists {
+		return nil, errors.New("v2.CiliumNetworkPolicy does not exist in local cache")
+	}
+
+	serverRule, ok := serverRuleStore.(*types.SlimCNP)
+	if !ok {
+		return nil, errors.New("received object of unknown type from API server, expecting v2.CiliumNetworkPolicy")
+	}
+
+	return serverRule, nil
+}
+
+func watchForCNPStatusEvents(ciliumStore cache.Store) {
+	if !kvstoreEnabled() {
+		log.Info("kvstore disabled, not watching for CNPStatus events from kvstore")
+		return
+	}
+
+	mgr := &CNPStatusEventManager{
+		eventMap: make(map[string]chan *NodeStatusUpdate),
+	}
+
+restart:
+	log.Info("starting kvstore watcher for CNP NodeStatus events")
+	watcher := kvstore.Client().ListAndWatch("cnpStatusWatcher", k8s.CNPStatusesPath, 512)
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				log.Debugf("%s closed, restarting watch", watcher.String())
+				time.Sleep(500 * time.Millisecond)
+				goto restart
+			}
+
+			switch event.Typ {
+			// Deletion from the kvstore of node statuses is performed by
+			// each cilium-agent.
+			case kvstore.EventTypeListDone, kvstore.EventTypeDelete:
+			case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+				var cnpStatusUpdate cilium_v2.CiliumNetworkPolicyNodeStatus
+				err := json.Unmarshal(event.Value, &cnpStatusUpdate)
+				if err != nil {
+					log.WithFields(logrus.Fields{"kvstore-event": event.Typ.String(), "key": event.Key}).
+						WithError(err).Error("Not updating CNP Status; error unmarshaling data from key-value store")
+					continue
+				}
+
+				namespace, name, node, err := extractFieldsFromKey(event.Key)
+				if err != nil {
+					log.WithFields(logrus.Fields{"kvstore-event": event.Typ.String(), "key": event.Key}).
+						WithError(err).Error("Not updating CNP Status; error extracting fields from key")
+					continue
+				}
+
+				log.WithFields(logrus.Fields{
+					"name":      name,
+					"namespace": namespace,
+					"node":      node,
+					"key":       event.Key,
+					"type":      event.Typ,
+				}).Debug("received event from kvstore")
+
+				nameNamespace := fmt.Sprintf("%s/%s", namespace, name)
+				mgr.mutex.Lock()
+				ch, ok := mgr.eventMap[nameNamespace]
+				if !ok {
```

Comment:
@aanm would checking if the CNP for which I received the update is in the `cache.Store` resolve the potential race [here](https://github.com/cilium/cilium/pull/9384/files#r333750774) ? Let me know if you need me to elaborate / explain what I am thinking of - I had a bit of trouble putting it into words when thinking of this.